hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java

```diff
@@ -108,7 +108,8 @@ protected String tableName() {
   @Override
   public FileIO io() {
     if (fileIO == null) {
-      fileIO = new HadoopFileIO(conf);
+      // avoid refresh metadata because refresh() calls io() again
+      fileIO = CatalogUtil.loadFileIO(getCurrentMetadataNoRefresh(), conf);
```

Comment:
What do you think about moving this to the base class? Then we wouldn't need to add the "no refresh" method to get metadata and could access `current` directly. This PR also makes it far less likely that implementations will override `io`, since a new one can be plugged in easily through table properties.

Reply:
I thought about doing that, but (1) if it is put to base class, the logic still needs to be duplicated, because `HadoopTableOperations` does not extend that base class; (2) the method needs to take the Hadoop configuration object, and it seems like the base class tries to not assume an implementation must use Hadoop configuration.

Reply:
Good point about Configuration. We don't want to rely on it in the base class. Let me take a closer look at this problem, it's concerning that there's a loop.

Reply:
Yeah I am also just thinking about the loop issue. If someone creates a new table and then reuse the same `TableOperations` object then it will continue to use the default `FileIO` and it would be a problem. However, `refresh` calls `refreshFromMetadataLocation` and that in the end has to call `io` to get all the metadata.

Reply:
Yeah, and we have to instantiate `FileIO` before we load the metadata for a table, so we don't really know if we're going to get it right. Seems like we are trying to configure this in the wrong place. Let's talk about where this should go in the sync today.

Reply:
Because Iceberg catalog does not enforce the implementation detail of how table properties are store, an implementation can potentially write the table properties using the `io()`. Therefore if we try to determine the `io()` to use based on table property, there is always a chance for cyclic dependency in logic. `HadoopCatalog` has exactly this problem.

Here are 3 potential ways to solve this problem:

1. Create a new constructor for each `TableOperation` that accepts `FileIO` as an argument. If the constructor is used, then `FileIO` is set at construction time instead of the first time `io()` is called. A default implementation can be loaded based on namespace properties. Engines like Spark and Flink can load a configuration key from Hadoop config to load the FileIO outside the core logic. This requires code change at multiple places, including (1) add new constructors in existing table operations, and (2) add support in each engine separately.

2. Determine `FileIO` based on the warehouse path scheme. For example, `s3://` always use `S3FileIO`, `hdfs://` always use `HadoopFileIO`. However, it is easy to create a cyclic dependency issue, for example: `iceberg-aws` module depends on `iceberg-core`, so `HadoopCatalog` in `iceberg-core` cannot import `S3FileIO` in `iceberg-aws`. This means we need a registry that allows people to register custom IO mapping at runtime. This approach has a similar amount of work as approach 1, because we need code change in each existing catalog, and each engine needs to find a way to register `FIleIO` implementation to scheme mapping at JVM start time.

3. Load `FileIO` using Hadoop configuration. Because `HadoopFileIO` is always the default implementation, Hadoop configuration object is always passed in. So user can always just define custom implementation at `core-site.xml`. This creates a simple solution with no dependency concern. However, this is not an elegant option because ideally we would like to load all Iceberg classes using Iceberg-based configs such as table properties.

Reply:
One quick thing to note: we want to avoid increasing dependence on a Hadoop configuration. It's fine to pass one where it is required, but we should always make sure there is an alternative and should generally avoid keeping properties there.

Thanks for writing up the options. Sounds like we have options for configuring a `FileIO`:
1. Pass `FileIO` in from the catalog and use catalog config to initialize it
2. Instantiate `FileIO` based on table location or metadata path just before using it
3. Use config from the environment, like a Hadoop FileSystem

I think the best choice is #1, catalog-level configuration.

We can't use table-level config because that creates a situation where it isn't available to load a table. Working around this requires deciding which FileIO to load based on different information at different times (location for create, metadata location for load), and would also make supplying a custom `FileIO` implementation in configuration difficult.

Environment-level config doesn't fit with the current model for customizing behavior, where everything is injected through `Catalog` and `Table`. Plus, Iceberg has no environment config like Hadoop `Configuration` and I don't think it is a good idea to add it. I think the most sensible thing is to maintain the chain of configuration: application passes config to catalogs, catalogs pass config to tables.

I don't think it would be too difficult to change over to `FileIO` passed in from the catalog, but this would mean not basing the implementation on table path. We wouldn't know to choose between `S3FileIO` or `HadoopFileIO` for a table and would have to use one for all tables from a catalog, or build a `FileIO` that knows when to choose between the two. I was already thinking that we would need a delegating implementation, since `S3FileIO` can't handle non-S3 paths. That should be okay.